docs/jax-101/02-jitting.ipynb

```diff
@@ -0,0 +1,570 @@
+    "\n",
+    "For `jax.jit`, the default level is `ShapedArray` -- that is, each tracer has a concrete shape (which we're allowed to condition on), but no concrete value. This allows the compiled function to work on all possible inputs with the same shape -- the standard use case in machine learning. However, because the tracers have no concrete value, if we attempt to condition on one, we get the error above.\n",
+    "\n",
+    "In `jax.grad`, the constraints are more relaxed, so you can do more. If you compose several transformations, however, you must satisfy the constraints of the most strict one. So, if you `jit(grad(f))`, `f` mustn't condition on value. For more detail on the interaction between Python control flow and JAX, see [this section](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Control-Flow) of the JAX docs.\n",
+    "\n",
+    "One way to deal with this problem is to rewrite the code to avoid conditionals on value. Another is to use special [control flow operators](https://jax.readthedocs.io/en/latest/jax.lax.html#control-flow-operators) like `jax.lax.cond`. However, sometimes that is impossible. In that case, you can consider jitting only part of the function. For example, if the most computationally expensive part of the function is inside the loop, we can jit just that inner part (though make sure to check the next section on caching to avoid shooting yourself in the foot):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {
+    "id": "OeR8hF-NHAML",
+    "outputId": "2b15ebbc-0fd8-4de1-e37f-a25fb5d4a522"
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "DeviceArray(30, dtype=int32)"
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {
+      "tags": []
+     },
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# While loop conditioned on x and n with a jitted body.\n",
+    "\n",
+    "@jax.jit\n",
+    "def loop_body(prev_i):\n",
+    "  return prev_i + 1\n",
+    "\n",
+    "def g_inner_jitted(x, n):\n",
+    "  i = 0\n",
+    "  while i < n:\n",
+    "    i = loop_body(i)\n",
+    "  return x + i\n",
+    "\n",
+    "g_inner_jitted(10, 20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "5XUT2acoHBz-"
+   },
+   "source": [
+    "If we really need to jit a function that has a condition on the value of an input, we can tell JAX to help itself to a less abstract tracer for a particular input by specifing `static_argnums`. The cost of this is that the resulting jaxpr is less flexible, so JAX will have to re-compile the function for every new value of the specified input. It is only a good strategy if the function is guaranteed to get limited different values."
```

Comment:
Suggestion:

- Another one, based on https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html#to-jit-or-not-to-jit 
which you can totally ignore, if you have an established way to write JIT/`jit` 😆 

```suggestion
    "If we really need to JIT a function that has a condition on the value of an input, we can tell JAX to help itself to a less abstract tracer for a particular input by specifing `static_argnums`. The cost of this is that the resulting jaxpr is less flexible, so JAX will have to re-compile the function for every new value of the specified input. It is only a good strategy if the function is guaranteed to get limited different values."
```